Klebert-Engineering/erdblick
# Viewport-Driven Tile Rendering

Viewport-Driven Tile Rendering
==============================

This PR introduces viewport-based map data fetching for Erdblick, now making full use of the mapget `/tiles` streaming response. Closes #19. Requires [#Klebert-Engineering/mapget#](https://github.com/Klebert-Engineering/mapget/pull/30).

![ezgif-3-bb4143198d](https://github.com/Klebert-Engineering/erdblick/assets/8723581/edd0d0e6-bed3-48d1-9346-99940c56171e)

How to test (KE)?
-----------------

The viewport-driven updates are currently best tested with a GeoJSON datasource based on a directory of exported NDS.Classic tiles. Step-by-step:

1. Download the tiles from [here](https://klebertengineering-my.sharepoint.com/:u:/g/personal/j_birkner_klebert-engineering_de/EaAGW52EEjBBnq3HneGaWZEByxEwQT8JZiOgHN18QxlCqA?e=UJwuRZ). Use the  "Routing JSON Export Password" from KE-Shared to unpack the archive.
2. In Erdblick, check out this branch (`feature/viewport-driven-updates`) and run `./ci/20_linux_rebuild.bash`.
3. In your local clone of `mapget-datasource-live`, check out `release/0.2.0`.
4. With you local build of `mapget-datasource-live`, run `./livesource-server geojsonsrc --source-dir path/to/routing-json-export --with-attr-layers false`. This will launch the livesource in geojson mode on port 8889. The `--with-attr-layers false` makes the performance more akin to what can be expected with staged loading.
5. With mapviewer or mapget project, run `./mapget serve -p 8089 -d 127.0.0.1:8889 -w "path/to/erdblick/build"`.

Follow-up work
--------------

1. I added a minimal "Focus" button to map-layer entries in the maps panel. This will only work if the layer info provides a tile-id list in its `coverage` field.
2. When many data sources are connected, the frontend seems to stall occasionally when many tiles are visible. Need to change how visualisation vs parsing tasks are prioritized when many, many tiles are coming in.
3. There is no usage of WebWorkers as of yet.
6. Currently, there is a lot of performance timing console output. This should be toggle-able.
7. I have activated the Cesium OSM layer, but it should be toggle-able.
8. We should prioritize the work on staged tile loading.
9. The dot-indicators for tiles-to-be-loaded are currently not toggle-able in any way.
10. You can use wheel-click-drag to turn the camera. But I have the feeling, that the view-direction-based tile prioritization is not working. This is probably not super apparent, since the prioritization by viewport center *is* working well.
